rust/eos-rs/src/api/dungeon_mode/dungeon_struct.rs

```diff
@@ -0,0 +1,990 @@
+        &self.1
+    }
+
+    /// Returns a mutable reference to the inner dungeon struct.
+    pub fn inner_mut(&mut self) -> &mut Dungeon<&'a mut ffi::dungeon> {
+        &mut self.1
+    }
+
+    /// Seems to initialize the dungeon struct from specified dungeon data.
+    ///
+    /// The signature will be updated once we know more about this function.
+    pub fn initialize_dungeon(
+        &mut self,
+        dungeon_data: &mut ffi::undefined,
+        dungeon: &mut ffi::dungeon,
+    ) -> i32 {
+        // SAFETY:We hold a valid mutable reference to the global dungeon struct.
+        unsafe { ffi::InitializeDungeon(dungeon_data as *mut _, dungeon as *mut _) }
+    }
+
+    /// Returns an abstraction over the game's builtin dungeon generator. This
+    /// struct implements [`dungeon_generator::DungeonFloorGeneration`], which is the
+    /// recommended way to work with it, see the documentation of [`dungeon_generator`].
+    ///
+    /// # Note
+    /// Note that the builtin dungeon generator works on the global dungeon struct directly.
+    pub fn get_builtin_dungeon_generator(
+        &'a mut self,
+    ) -> dungeon_generator::game_builtin::GlobalDungeonStructureGenerator<'a> {
+        dungeon_generator::game_builtin::GlobalDungeonStructureGenerator(self.0.clone(), self)
+    }
+
+    /// Generates a dungeon floor.
+    ///
+    /// If not changed by a patch, this function will use the game's default built in generator
+    /// and generate the floor based on the current global configuration for the floor.
+    ///
+    /// For more granular control over the floor generation, you can get an abstraction over
+    /// the builtin generator with [`Self::get_builtin_dungeon_generator`].
+    ///
+    pub fn generate_floor(&'a mut self) {
+        self.get_builtin_dungeon_generator()
+            .generate_floor_internal();
+    }
+
+    /// Gets the floor type. Returns None if the global dungeon struct contains invalid data.
+    pub fn get_floor_type(&self) -> Option<FloorType> {
+        // SAFETY:We hold a valid mutable reference to the global dungeon struct.
+        unsafe { ffi::GetFloorType() }.try_into().ok()
+    }
+
+    /// Checks if the current fixed floor is the "substitute room" (Fixed Room ID 0x6E).
+    pub fn is_substitute_room(&self) -> bool {
+        // SAFETY:We hold a valid mutable reference to the global dungeon struct.
+        unsafe { ffi::FixedRoomIsSubstituteRoom() > 0 }
+    }
+
+    /// Checks if the current dungeon floor number is even.
+    /// This is probably, among other things(?), used to determine whether male or female monsters
+    /// should be spawned.
+    /// Has a special check to return false for Labyrinth Cave B10F (the Gabite boss fight).
+    pub fn is_even_floor(&self) -> bool {
+        // SAFETY:We hold a valid mutable reference to the global dungeon struct.
+        unsafe { ffi::FloorNumberIsEven() > 0 }
+    }
+
+    /// Returns the tile at the given coordinates.
```

Comment:
We should specify what it does when the tile is out-of-bound. I've looked at the game code, and it seems to return a (shared) default tile.